bionic/deriver.py

```diff
@@ -180,24 +243,77 @@ def _prevalidate_dnodes(self, dnodes):
     def _get_base_dnodes(self):
         """
         Returns the list of descriptor nodes needed to compute all user-defined entities
-        and internal entities. (At the time of writing, this includes every
-        valid descriptor, but in the future there will be an infinite number
-        of valid descriptors.)
+        and internal entities.
         """
 
         return [
             entity_dnode_from_descriptor(entity_name)
-            for entity_name in self._flow_state.providers_by_name.keys()
+            for entity_name in self._flow_state.entity_defs_by_name.keys()
         ]
 
+    def _obtain_provider_for_dnode(self, dnode):
+        """
+        Returns a Provider object for a given descriptor node -- either by finding
+        a statically precomputed provider, or automatically generating one.
+        """
+
+        assert self._static_providers_by_dnode is not None
+        if dnode in self._static_providers_by_dnode:
+            return self._static_providers_by_dnode[dnode]
+
+        if isinstance(dnode, ast.EntityNode):
+            entity_name = dnode.to_entity_name()
+            if entity_name in self._flow_state.entity_defs_by_name:
+                message = f"""
+                Unexpected failed to find a static provider for defined entity
+                {entity_name!r};
+                this should be impossible!
+                """
+                raise AssertionError(oneline(message))
+
+            raise UndefinedEntityError.for_name(entity_name)
+
+        elif isinstance(dnode, ast.TupleNode):
+            return TupleConstructionProvider(dnode)
+
+        else:
+            raise AssertionError(
+                f"Unexpected dnode type {type(dnode)!r} for dnode {dnode!r}"
+            )
+
+    def _obtain_entity_def_for_dnode(self, dnode):
+        """
+        Returns an EntityDefinition object for a given descriptor node. If the node is
+        an entity, this returns the definition specified by the user; if the node is a
+        tuple, we generate a synthetic definition.
+        """
+
+        if isinstance(dnode, ast.EntityNode):
+            return self._flow_state.get_entity_def(dnode.to_descriptor())
+
+        elif isinstance(dnode, ast.TupleNode):
+            # TODO Since we're using this to describe things something that's not an
+            # entity, we should rename `EntityDefinition` to something more general.
+            return EntityDefinition(
+                name=dnode.to_descriptor(),
+                protocol=TupleProtocol(len(dnode.children)),
+                doc=f"A Python tuple with {len(dnode.children)} values.",
+                can_memoize=True,
```

Comment:
Yes, I think this the same issue I mentioned here: https://github.com/square/bionic/pull/150#issuecomment-633046597. I don't think this will be a problem until we actually merge that other PR; until then this code shouldn't get used.

As to how we can resolve it: Serializing tuples is tricky, because we'll need access to the appropriate protocol for each of the tuple elements as well. And if we don't serialize the tuple, we need to memoize it long enough to compute each of its downstream descriptor values. That's why I'm inclined to fix this by keeping the memoization, but having a policy of discarding memoized values after all their downstream tasks have been completed.

If this makes sense to you, I can certainly add a comment explaining this.

Reply:
Right, it makes sense to memoize it for now. 

I'm not very sure how you are defining downstream tasks here though. Is it when the `TaskCompletionRunner` completes a run? Or are you going to track where all is this value required and drop it as soon as it is no longer needed even in the middle of `TaskCompletionRunner.run()`? If it's the former, then we might still keep a lot of data in memory if a run computes a lot of big objects. If it's the latter, you will have to add a lot of tracking logic to make it work. 

My proposal is to not memoize or persist tuples. We compute the tuple whenever someone needs it inside compute function. That way, tuple can rely on children's caching behavior. The only problem is that we require an entity to be either memoized or persisted so you'll have to add support for it too. 

Any approach you take, you'll have to add some kind of additional logic that doesn't exist yet to support the approach. I'll be fine with you leaving a comment explaining all this.

Reply:
I was thinking of the latter. There would definitely be some additional tracking logic but hopefully not too much. (I think it might just be a matter of moving a task state from `COMPLETE` to "untracked" when nothing else depends on it.) I haven't totally thought it through but it seems like the best tradeoff so far, since it resolves this particular issue and generally reduces unnecessary memory usage.

Alternatively, if you "compute the tuple whenever someone needs it", doesn't that mean you'll end up computing it once for each component of the tuple? For example, in the example ML workflow, 'train_frame` and `test_frame` are both upstream of `precision_recall_frame`. It seems like we would have to compute the tuple `train_frame, test_frame` once in order to compute and persist `train_frame`, and then a second time to compute and persist `test_frame`. Is that right?

Reply:
We discussed this elsewhere, leading to [this comment](https://github.com/square/bionic/pull/150#issuecomment-640723925) in another PR.